python-checks/src/main/java/org/sonar/python/checks/DuplicatedMethodImplementationCheck.java

```diff
@@ -0,0 +1,103 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.python.api.tree.ClassDef;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.StatementList;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.IssueLocation;
+
+@Rule(key = "S4144")
+public class DuplicatedMethodImplementationCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Update this function so that its implementation is not identical to %s on line %s.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CLASSDEF, ctx -> {
+      ClassDef classDef = (ClassDef) ctx.syntaxNode();
+      MethodVisitor methodVisitor = new MethodVisitor();
+      classDef.body().accept(methodVisitor);
+
+      for (int i = 1; i < methodVisitor.methods.size(); i++) {
+        checkMethods(methodVisitor.methods.get(i), methodVisitor.methods, i, ctx);
+      }
+    });
+  }
+
+  private static void checkMethods(FunctionDef suspiciousMethod, List<FunctionDef> methods, int index, SubscriptionContext ctx) {
+    StatementList suspiciousBody = suspiciousMethod.body();
+    if (isException(suspiciousMethod)) {
+      return;
+    }
+    for (int j = 0; j < index; j++) {
+      FunctionDef originalMethod = methods.get(j);
+      Tree originalBody = originalMethod.body();
+      if (CheckUtils.areEquivalent(originalBody, suspiciousBody)) {
+        int line = originalMethod.name().firstToken().line();
+        String message = String.format(MESSAGE, originalMethod.name().name(), line);
+        ctx.addIssue(suspiciousMethod.name(), message).secondary(IssueLocation.preciseLocation(originalMethod.name(), "Original"));
+        break;
+      }
+    }
+  }
+
+  private static boolean isException(FunctionDef suspiciousMethod) {
+    boolean hasDocString = suspiciousMethod.docstring() != null;
+    StatementList suspiciousBody = suspiciousMethod.body();
+    int nbActualStatements = hasDocString ? suspiciousBody.statements().size() - 1 : suspiciousBody.statements().size();
+    if (nbActualStatements == 0 || isOnASingleLine(suspiciousBody, hasDocString)) {
+      return true;
+    }
+    return nbActualStatements == 1 && suspiciousBody.statements().get(suspiciousBody.statements().size() - 1).is(Tree.Kind.RAISE_STMT);
```

Comment:
Let's create a variable for `suspiciousBody.statements()` instead of duplicating the expression 4 times.